File: src/main/java/com/jennifertellez/library/service/BookService.java
package com.jennifertellez.library.service;

import com.jennifertellez.library.dto.BookResponse;
import com.jennifertellez.library.dto.CreateBookRequest;
import com.jennifertellez.library.dto.PageResponse;
import com.jennifertellez.library.dto.UpdateBookRequest;
import com.jennifertellez.library.model.ReadingStatus;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface BookService {

    BookResponse createBook(CreateBookRequest request);

    BookResponse getBookById(Long id);

    List<BookResponse> getAllBooks();

    List<BookResponse> getBooksByStatus(ReadingStatus status);

    BookResponse updateBook(Long id, UpdateBookRequest request);

    void deleteBook(Long id);

    List<BookResponse> searchBooks(String searchTerm);

    PageResponse<BookResponse> getAllBooks(Pageable pageable);

    PageResponse<BookResponse> searchBooks( String searchTerm, Pageable pageable);

    PageResponse<BookResponse> getBooksByStatus(ReadingStatus status, Pageable pageable);

    PageResponse<BookResponse> getBooksByAuthor(String author, Pageable pageable);

}

File: library-frontend/src/services/bookService.ts
import { api } from './api';
import { ReadingStatus } from '../types/book';
import type {
  Book,
  CreateBookRequest,
  UpdateBookRequest,
  PageResponse
} from '../types/book';

export const bookService = {
  // Get all books with pagination
  getAllBooks: async (page = 0, size = 20, sortBy = 'title', direction = 'ASC') => {
    const response = await api.get<PageResponse<Book>>('/books', {
      params: { page, size, sortBy, direction }
    });
    return response.data;
  },

  // Get book by ID
  getBookById: async (id: number) => {
    const response = await api.get<Book>(`/books/${id}`);
    return response.data;
  },

  // Create book
  createBook: async (book: CreateBookRequest) => {
    const response = await api.post<Book>('/books', book);
    return response.data;
  },

  // Create book from ISBN
  createBookFromIsbn: async (isbn: string) => {
    const response = await api.post<Book>(`/books/isbn/${isbn}`);
    return response.data;
  },

  // Update book
  updateBook: async (id: number, book: UpdateBookRequest) => {
    const response = await api.put<Book>(`/books/${id}`, book);
    return response.data;
  },

  // Delete book
  deleteBook: async (id: number) => {
    await api.delete(`/books/${id}`);
  },

  // Search books
  searchBooks: async (term: string, page = 0, size = 20) => {
    const response = await api.get<PageResponse<Book>>('/books/search', {
      params: { term, page, size }
    });
    return response.data;
  },

  // Get books by status
  getBooksByStatus: async (status: string, page = 0, size = 20) => {
    const response = await api.get<PageResponse<Book>>('/books', {
      params: { status, page, size }
    });
    return response.data;
  },

  // Get reading statistics
  getStats: async () => {
    const response = await api.get<ReadingStats>('/books/stats');
    return response.data;
  },
};